common/src/main/java/io/novafoundation/nova/common/utils/recyclerView/expandable/ExpandableItemDecoration.kt

```diff
@@ -0,0 +1,98 @@
+
+private data class ItemWithViewHolder(val position: Int, val item: ExpandableBaseItem, val viewHolder: ViewHolder?)
+
+abstract class ExpandableItemDecoration(
+    private val adapter: ExpandableAdapter,
+    private val animator: ExpandableAnimator
+) : RecyclerView.ItemDecoration() {
+
+    abstract fun onDrawGroup(
+        canvas: Canvas,
+        animationState: ExpandableAnimationItemState,
+        recyclerView: RecyclerView,
+        parentItem: ExpandableParentItem,
+        parent: ViewHolder?,
+        children: List<ViewHolder>
+    )
+
+    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
+    }
+
+    override fun onDraw(canvas: Canvas, recyclerView: RecyclerView, state: RecyclerView.State) {
+        val items = getParentAndChildren(recyclerView)
+        for ((parentItem, children) in items) {
+            val animationState = animator.getStateForPosition(parentItem.position) ?: continue
+            val childViewHolders = children.mapNotNull { it.viewHolder }
+            onDrawGroup(canvas, animationState, recyclerView, parentItem.item as ExpandableParentItem, parentItem.viewHolder, childViewHolders)
+        }
+    }
+
+    private fun getParentAndChildren(recyclerView: RecyclerView): Map<ItemWithViewHolder, List<ItemWithViewHolder>> {
+        // Searching all view holders in recycler view and match them with adapter items
+        val items = recyclerView.children.toList()
+            .mapNotNull {
+                val viewHolder = recyclerView.getChildViewHolder(it)
+                val expandableViewHolder = viewHolder as? ExpandableBaseViewHolder<*> ?: return@mapNotNull null
+                val item = expandableViewHolder.expandableItem ?: return@mapNotNull null
+                ItemWithViewHolder(viewHolder.bindingAdapterPosition, item, viewHolder)
+            }
+
+        // Grouping view holders by parents
+        val parentsWithChildren = mutableMapOf<ItemWithViewHolder, MutableList<ItemWithViewHolder>>()
+
+        val parents = items.filter { it.item is ExpandableParentItem }.associateBy { it.item.getId() }
+        val children = items.filter { it.item is ExpandableChildItem }
+        parents.values.forEach { parentsWithChildren[it] = mutableListOf() }
+
+        children.forEach { child ->
+            val item = child.item as ExpandableChildItem
+            val parent = parents[item.groupId] ?: getParentForItem(recyclerView, item) ?: return@forEach
+            val parentChildren = parentsWithChildren[parent] ?: mutableListOf()
+            parentChildren.add(child)
+            parentsWithChildren[parent] = parentChildren
+        }
+
+        return parentsWithChildren
+    }
+
+    private fun getParentForItem(recyclerView: RecyclerView, item: ExpandableChildItem): ItemWithViewHolder? {
+        val positionInAdapter = adapter.getItems().indexOfFirstOrNull { it.getId() == item.groupId } ?: return null
```

Comment:
Seems like we are looping trough every adapter iteam here and the `getParentForItem` itself is called inside the loop (` children.forEach`. This might introduce performance issue. Can you verify how the scrolling and animation works with large datasets? E.g. many expanded parents with hide zero balances turned off